### PR TITLE
feat(lockfiles): support Yarn 4 patches

### LINKED ISSUE #6450: [turborepo] Yarn v4 patch causes `turbo prune` to panic, no locator found
### What version of Turborepo are you using?

1.10.16

### What package manager are you using / does the bug impact?

Yarn v2/v3 (node_modules linker only)

### What operating system are you using?

Mac

### Describe the Bug

```txt
yarn turbo prune --scope MobileApp
Generating pruned monorepo for MobileApp in /Users/username/projectname
 - Added @project/client-shared
 - Added @project/core
 - Added @project/graphql
 - Added @project/openapi
 - Added @project/tsconfig
 - Added @project/types
 - Added MobileApp
 - Added eslint-config-project
thread 'main' panicked at crates/turborepo-lockfiles/src/berry/mod.rs:456:44:
No locator found for @openapi-codegen/typescript@npm:^6.2.0
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Where I have the following (simplified) `package.json`:

root:

```json
{
  "packageManager": "yarn@4.0.2",
  "resolutions": {
    "@openapi-codegen/typescript@^6.2.0": "patch:@openapi-codegen/typescript@npm%3A6.2.0#./.yarn/patches/@openapi-codegen-typescript-npm-6.2.0-c197461d23.patch"
  }
}
```

`@project/openapi`

```json
{
  "name": "@project/openapi",
  "devDependencies": {
    "@openapi-codegen/typescript": "^6.2.0"
  }
}
```

Partial `yarn.lock`

```yml
"@openapi-codegen/typescript@npm:6.2.0":
  version: 6.2.0
  resolution: "@openapi-codegen/typescript@npm:6.2.0"
  dependencies:
    case: "npm:^1.6.3"
    lodash: "npm:^4.17.21"
    openapi3-ts: "npm:^2.0.1"
    pluralize: "npm:^8.0.0"
    tslib: "npm:^2.3.1"
    tsutils: "npm:^3.21.0"
    typescript: "npm:4.8.2"
  checksum: cbf9d6cbd1489cd47e79cb28aed5a2e3f882aca58d30f3a5a83ad5ca943af656df7c4e780abf49de78fb316ebcbd7cb79fd9b64b7d7183d14d76a6985331d4df
  languageName: node
  linkType: hard

"@openapi-codegen/typescript@patch:@openapi-codegen/typescript@npm%3A6.2.0#./.yarn/patches/@openapi-codegen-typescript-npm-6.2.0-c197461d23.patch::locator=project%40workspace%3A.":
  version: 6.2.0
  resolution: "@openapi-codegen/typescript@patch:@openapi-codegen/typescript@npm%3A6.2.0#./.yarn/patches/@openapi-codegen-typescript-npm-6.2.0-c197461d23.patch::version=6.2.0&hash=ed8498&locator=project%40workspace%3A."
  dependencies:
    case: "npm:^1.6.3"
    lodash: "npm:^4.17.21"
    openapi3-ts: "npm:^2.0.1"
    pluralize: "npm:^8.0.0"
    tslib: "npm:^2.3.1"
    tsutils: "npm:^3.21.0"
    typescript: "npm:4.8.2"
  checksum: ec10908cbf51fb704066ba591f1294f389ab4606063ad76359061d3db35ad6d045bc88e8b09c0469dffe9a61f53d111533525054b1ec8f85febc17f005befb61
  languageName: node
  linkType: hard
  
"@project/openapi@workspace:*, @project/openapi@workspace:packages/openapi":
  version: 0.0.0-use.local
  resolution: "@project/openapi@workspace:packages/openapi"
  dependencies:
    "@openapi-codegen/typescript": "npm:^6.2.0"
    "@project/tsconfig": "workspace:*"
  languageName: unknown
  linkType: soft

```

### Expected Behavior

Should not panic.

### To Reproduce

1. (I started with a Yarn 3 patch then upgraded to Yarn 4. You might be able to create a patch on Yarn 4 immediately rather than 3 -> 4, not sure).
2. Have a patch exist for a dependency in one of the packages in the prune
3. Run `yarn prune`

I can email you a link to a repo if needed.

### Reproduction Repo

_No response_

<sub>[TURBO-1719](https://linear.app/vercel/issue/TURBO-1719/[turborepo]-yarn-v4-patch-causes-turbo-prune-to-panic-no-locator-found)</sub>

### PR DESCRIPTION
**Pull request by chris-olszewski**
### Description

Yarn 4 changed how patches can get encoded in the lockfile and `resolutions` field in the root `package.json`. Primarily the changes are with direct dependencies that are patched. These now will appear in the `package.json` and the workspace's `dependency`map in the lockfile entry.

This PR alters our behavior to now properly handle these direct dependencies.

Closes #6450

### Testing Instructions

Added new unit tests with Yarn 4 fixtures. Tested out pruning manually by creating a test repo with direct and indirect patches:

#### Setup
```
npx create-turbo@latest yarn4 yarn
cd yarn4
echo "nodeLinker: node-modules" >> .yarnrc.yml
yarn set version 4
cd apps/docs
yarn add is-odd
yarn patch is-odd
# follow patch commands to patch is-odd
cd ../web
yarn add is-even
cd ../..
```

#### Testing
Test that pruning both `web` and `docs` prune properly:
```
[0 olszewski@chriss-mbp] /tmp/yarn4 $ turbo_dev prune web
Generating pruned monorepo for web in /private/tmp/yarn4/out
 - Added @repo/eslint-config
 - Added @repo/typescript-config
 - Added @repo/ui
 - Added web
[0 olszewski@chriss-mbp] /tmp/yarn4 $ cd out 
[0 olszewski@chriss-mbp] /tmp/yarn4/out $ yarn install --immutable
```

Closes TURBO-2022

### CODE REVIEW

File: crates/turborepo-lockfiles/src/berry/mod.rs [hunk 1]
```diff
@@ -461,103 +491,8 @@ impl Lockfile for BerryLockfile {
...
-            patches,
-            // We clone the following structures without any alterations and
-            // rely on resolutions being correctly pruned.
-            locator_package: self.locator_package.clone(),
-            resolver: self.resolver.clone(),
-            extensions: self.extensions.clone(),
-            overrides: self.overrides.clone(),
-            workspace_path_to_locator: self.workspace_path_to_locator.clone(),
-        }))
+        let subgraph = self.subgraph(workspace_packages, packages)?;
```
Comment (chris-olszewski) [inline crates/turborepo-lockfiles/src/berry/mod.rs:494]: This is an oopsie from 6 months ago where we ended up having duplicate definitions for `subgraph` one on `BerryLockfile` and one on the `Lockfile` implementation. This unifies them by forwarding the trait implementation to the type one.

File: crates/turborepo-lockfiles/src/berry/mod.rs [hunk 2]
```diff
@@ -249,7 +259,7 @@ impl BerryLockfile {
 
     /// Produces a new lockfile containing only the given workspaces and
     /// packages
-    pub fn subgraph(
+    fn subgraph(
```
Comment (chris-olszewski) [inline crates/turborepo-lockfiles/src/berry/mod.rs:262]: Make this private to ensure external callers go through the trait and don't call this directly.

File: crates/turborepo-lockfiles/src/berry/mod.rs [hunk 3]
```diff
@@ -756,7 +691,7 @@ mod test {
 
         assert_eq!(
             &lockfile.extensions,
-            &(["@babel/types@npm:^7.8.3", "lodash@npm:4.17.21"]
+            &(["@babel/types@npm:^7.8.3"]
```
Comment (chris-olszewski) [inline crates/turborepo-lockfiles/src/berry/mod.rs:694]: This test asserted some buggy behavior as it identified `lodash` as being an extension when it was really a user patch.

What we truly care about is that this `lodash` package gets included in the closure. This is asserted by the `test_closure_with_patch` test.

**Review by chris-olszewski — COMMENTED** [review]
Reviewer's Guide

**Review by arlyon — APPROVED** [review]
Surprisingly few changes :) small Q for my understanding but otherwise good

File: crates/turborepo-lockfiles/src/berry/mod.rs [hunk 4]
```diff
@@ -1048,4 +983,124 @@ mod test {
...
+                &["packages/b".into(), "packages/c".into()],
+                &[expected_key.into(), "is-number@npm:6.0.0".into()],
+            )
+            .unwrap();
+
+        let sublockfile = subgraph.lockfile().unwrap();
+
+        // Should contain both patched dependency and original
+        assert!(sublockfile.packages.contains_key("is-odd@npm:3.0.1"));
+        assert!(sublockfile.packages.contains_key(
```
[the remark this answers is not in the document]
Comment (chris-olszewski) [reply crates/turborepo-lockfiles/src/berry/mod.rs:1053]: At the moment no, but that is something we probably should do. Doing it properly will require a little bit of work to make it performant as currently we require building a whole pruned lockfile in order to know what patches affect a set of packages.

Currently the best solution would be to include `.yarn/patches` in the `globalDependencies`.

**Pull request merged**